syntax = "proto3";

package TW.Zilliqa.Proto;
option java_package = "wallet.core.jni.proto";

message Transaction {
    message Transfer {
        // Amount to send (256-bit number)
        bytes amount           = 1;
    }

    message Raw {
        // Amount to send (256-bit number)
        bytes amount           = 1;

        // Smart contract code
        bytes code             = 2;

        // String-ified JSON object specifying the transition parameter
        bytes data             = 3;
    }

    oneof message_oneof {
        Transfer transfer = 1;
        Raw raw_transaction = 2;
    }
}

// Input data necessary to create a signed transaction.
message SigningInput {
    // Transaction version
    uint32 version         = 1;

    // Nonce
    uint64 nonce           = 2;

    // Recipient's address.
    string to              = 3;

    // GasPrice (256-bit number)
    bytes gas_price        = 4;

    // GasLimit
    uint64 gas_limit       = 5;

    // Private Key
    bytes private_key      = 6;

    Transaction transaction = 7;
}

// Transaction signing output.
message SigningOutput {
    // Signed signature bytes.
    bytes signature = 1;

    // JSON transaction with signature
    string json = 2;
}
